File: tools/pass622_codex_ghidra_s2c_export/ghidra_export_s2c_receive_path.java
// Pass622 Ghidra Java exporter for native S2C receive/world-handshake path.
// Run inside Ghidra headless as a fallback when Jython scripting is unavailable.

import ghidra.app.script.GhidraScript;
import ghidra.app.decompiler.DecompInterface;
import ghidra.app.decompiler.DecompileResults;
import ghidra.program.model.address.Address;
import ghidra.program.model.listing.Function;
import ghidra.program.model.listing.FunctionManager;
import ghidra.program.model.listing.Instruction;
import ghidra.program.model.listing.InstructionIterator;
import ghidra.program.model.pcode.PcodeOp;
import ghidra.program.model.symbol.Reference;
import ghidra.program.model.symbol.ReferenceIterator;
import ghidra.program.model.symbol.Symbol;
import ghidra.program.model.symbol.SymbolIterator;

import java.io.File;
import java.io.FileWriter;
import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.HashSet;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.regex.Pattern;

public class ghidra_export_s2c_receive_path extends GhidraScript {
    private static final String[] NETWORK_NAMES = new String[] {
        "recv", "WSARecv", "recvfrom", "send", "WSASend", "closesocket",
        "connect", "select", "ioctlsocket", "InternetReadFile", "ReadFile"
    };
    private static final String[] VM_ANCHORS = new String[] {
        "0x11B562BD", "0x11B5630F", "0x11B5932F", "0x11B57796",
        "0x11B55DF6", "0x11B54E6F", "0x11B566B4", "0x11B56C63"
    };
    private static final int MAX_DEPTH = 3;

    private FunctionManager fm;
    private DecompInterface decomp;

    public void run() throws Exception {
        String[] args = getScriptArgs();
        String outPath = args.length > 0 ? args[0] : "C:\\AionTools\\aion_decoder_agent\\outbox\\pass622_ghidra_s2c_receive_exports";
        File outDir = new File(outPath);
        outDir.mkdirs();
        fm = currentProgram.getFunctionManager();
        decomp = new DecompInterface();
        decomp.openProgram(currentProgram);

        Map<String, Function> seeds = new LinkedHashMap<String, Function>();
        List<String> importRows = new ArrayList<String>();
        importRows.add("api,symbol_address,from_address,function,function_entry");

        for (String name : NETWORK_NAMES) {
            SymbolIterator syms = currentProgram.getSymbolTable().getSymbols(name);
            while (syms.hasNext()) {
                Symbol sym = syms.next();
                ReferenceIterator refs = currentProgram.getReferenceManager().getReferencesTo(sym.getAddress());
                while (refs.hasNext()) {
                    Reference ref = refs.next();
                    Function fn = functionAtOrContaining(ref.getFromAddress());
                    String fnName = fn == null ? "" : fn.getName();
                    String fnEntry = fn == null ? "" : addrText(fn.getEntryPoint());
                    importRows.add(csv(name) + "," + csv(addrText(sym.getAddress())) + "," + csv(addrText(ref.getFromAddress())) + "," + csv(fnName) + "," + csv(fnEntry));
                    if (fn != null) seeds.put(fnEntry, fn);
                }
            }
        }

        for (String anchor : VM_ANCHORS) {
            try {
                Function fn = functionAtOrContaining(toAddr(anchor));
                if (fn != null) seeds.put(addrText(fn.getEntryPoint()), fn);
            } catch (Exception ignored) {}
        }

        Map<String, Integer> callers = walk(seeds, true);
        Map<String, Integer> callees = walk(seeds, false);
        Map<String, Function> targets = new LinkedHashMap<String, Function>();
        targets.putAll(seeds);
        for (String k : callers.keySet()) targets.put(k, functionAtOrContaining(toAddr(k)));
        for (String k : callees.keySet()) targets.put(k, functionAtOrContaining(toAddr(k)));

        List<String> functionRows = new ArrayList<String>();
        List<String> writeRows = new ArrayList<String>();
        List<String> edgeRows = new ArrayList<String>();
        functionRows.add("entry,name,seed,caller_depth,callee_depth,vm_anchor_hits,disasm_file,pcode_file,decomp_file");
        writeRows.add("function,function_entry,pattern,line");
        edgeRows.add("from_entry,from_name,to_entry,to_name");

        for (String key : targets.keySet()) {
            Function fn = targets.get(key);
            if (fn == null) continue;
            String base = key.replace("0x", "") + "_" + fn.getName().replaceAll("[^A-Za-z0-9_]", "_");
            String disasm = disasmText(fn);
            String pcode = pcodeText(fn);
            String decompText = decompileText(fn);
            writeFile(new File(outDir, base + ".disasm.txt"), disasm);
            writeFile(new File(outDir, base + ".pcode.txt"), pcode);
            writeFile(new File(outDir, base + ".decomp.txt"), decompText);
            functionRows.add(csv(key) + "," + csv(fn.getName()) + "," + csv(Boolean.toString(seeds.containsKey(key))) + "," + csv(depthText(callers, key)) + "," + csv(depthText(callees, key)) + "," + csv(anchorHits(fn, disasm)) + "," + csv(base + ".disasm.txt") + "," + csv(base + ".pcode.txt") + "," + csv(base + ".decomp.txt"));
            for (Function callee : calleesOf(fn)) edgeRows.add(csv(key) + "," + csv(fn.getName()) + "," + csv(addrText(callee.getEntryPoint())) + "," + csv(callee.getName()));
            for (String line : (pcode + "\n" + disasm).split("\n")) {
                String up = line.toUpperCase();
                if (up.contains("STORE") || Pattern.compile("\\bMOV\\s+\\w+ PTR \\[").matcher(up).find()) {
                    String pat = (up.contains("QWORD") || up.contains(":8") || up.contains("(8)")) ? "8-byte-write" : "write";
                    if (up.contains("INT_XOR") || up.contains("INT_ADD") || up.contains("INT_MULT")) pat += "+key-arithmetic-nearby";
                    writeRows.add(csv(fn.getName()) + "," + csv(key) + "," + csv(pat) + "," + csv(line.length() > 240 ? line.substring(0, 240) : line));
                }
            }
        }

        writeFile(new File(outDir, "import_refs.csv"), join(importRows));
        writeFile(new File(outDir, "candidate_functions.csv"), join(functionRows));
        writeFile(new File(outDir, "call_edges.csv"), join(edgeRows));
        writeFile(new File(outDir, "write_hints.csv"), join(writeRows));
        writeFile(new File(outDir, "export_manifest.txt"), "program=" + currentProgram.getName() + "\nmax_depth=" + MAX_DEPTH + "\nfunctions=" + (functionRows.size() - 1) + "\nwrites=" + (writeRows.size() - 1) + "\n");
        println("Pass622 Java S2C receive export wrote " + (functionRows.size() - 1) + " functions to " + outPath);
    }

    private String addrText(Address a) { return "0x" + a.toString().toUpperCase(); }
    private Function functionAtOrContaining(Address a) { Function f = fm.getFunctionAt(a); return f != null ? f : fm.getFunctionContaining(a); }
    private String depthText(Map<String, Integer> m, String k) { return m.containsKey(k) ? Integer.toString(m.get(k)) : ""; }
    private String csv(String s) { if (s == null) s = ""; return "\"" + s.replace("\"", "\"\"") + "\""; }
    private String join(List<String> rows) { StringBuilder sb = new StringBuilder(); for (String r: rows) sb.append(r).append("\n"); return sb.toString(); }
    private void writeFile(File f, String s) throws Exception { FileWriter w = new FileWriter(f); w.write(s == null ? "" : s); w.close(); }

    private Map<String, Integer> walk(Map<String, Function> seeds, boolean callerDirection) {
        Map<String, Integer> seen = new HashMap<String, Integer>();
        ArrayDeque<Object[]> q = new ArrayDeque<Object[]>();
        for (Function f : seeds.values()) q.add(new Object[] {f, 0});
        while (!q.isEmpty()) {
            Object[] item = q.removeFirst();
            Function fn = (Function)item[0];
            int depth = ((Integer)item[1]).intValue();
            String key = addrText(fn.getEntryPoint());
            if (seen.containsKey(key) && seen.get(key) <= depth) continue;
            seen.put(key, depth);
            if (depth >= MAX_DEPTH) continue;
            List<Function> next = callerDirection ? callersOf(fn) : calleesOf(fn);
            for (Function n : next) q.add(new Object[] {n, depth + 1});
        }
        return seen;
    }

    private List<Function> callersOf(Function fn) {
        List<Function> out = new ArrayList<Function>();
        ReferenceIterator refs = currentProgram.getReferenceManager().getReferencesTo(fn.getEntryPoint());
        while (refs.hasNext()) { Function f = functionAtOrContaining(refs.next().getFromAddress()); if (f != null) out.add(f); }
        return out;
    }
    private List<Function> calleesOf(Function fn) {
        List<Function> out = new ArrayList<Function>();
        InstructionIterator it = currentProgram.getListing().getInstructions(fn.getBody(), true);
        while (it.hasNext()) {
            Instruction ins = it.next();
            for (Address a : ins.getFlows()) { Function f = functionAtOrContaining(a); if (f != null) out.add(f); }
        }
        return out;
    }
    private String disasmText(Function fn) {
        StringBuilder sb = new StringBuilder();
        InstructionIterator it = currentProgram.getListing().getInstructions(fn.getBody(), true);
        while (it.hasNext()) { Instruction ins = it.next(); sb.append(addrText(ins.getAddress())).append(" ").append(ins.toString()).append("\\n"); }
        return sb.toString();
    }
    private String pcodeText(Function fn) {
        StringBuilder sb = new StringBuilder();
        InstructionIterator it = currentProgram.getListing().getInstructions(fn.getBody(), true);
        while (it.hasNext()) { Instruction ins = it.next(); for (PcodeOp op : ins.getPcode()) sb.append(addrText(ins.getAddress())).append(" ").append(op.toString()).append("\\n"); }
        return sb.toString();
    }
    private String decompileText(Function fn) {
        try { DecompileResults r = decomp.decompileFunction(fn, 60, monitor); return r.decompileCompleted() ? r.getDecompiledFunction().getC() : ""; } catch (Exception e) { return "DECOMPILE_ERROR: " + e.toString(); }
    }
    private String anchorHits(Function fn, String disasm) {
        String up = (fn.getName() + " " + addrText(fn.getEntryPoint()) + " " + disasm).toUpperCase();
        List<String> hits = new ArrayList<String>();
        for (String a : VM_ANCHORS) if (up.contains(a.substring(2).toUpperCase())) hits.add(a);
        return String.join(";", hits);
    }
}

File: tools/pass622_codex_ghidra_s2c_export/ghidra_export_s2c_receive_path.py
# Pass622 Ghidra Jython exporter for native S2C receive/world-handshake path.
# Run inside Ghidra headless. Output is local-only and may contain p-code/disassembly.

from __future__ import print_function
import os
import re
import json

from ghidra.program.model.symbol import RefType
from ghidra.util.task import ConsoleTaskMonitor

NETWORK_NAMES = [
    "recv", "WSARecv", "recvfrom", "send", "WSASend", "closesocket",
    "connect", "select", "ioctlsocket", "InternetReadFile", "ReadFile"
]
VM_ANCHORS = [
    "0x11B562BD", "0x11B5630F", "0x11B5932F", "0x11B57796",
    "0x11B55DF6", "0x11B54E6F", "0x11B566B4", "0x11B56C63"
]
MAX_DEPTH = 3

args = getScriptArgs()
out_dir = args[0] if args else r"C:\AionTools\aion_decoder_agent\outbox\pass622_ghidra_s2c_receive_exports"
if not os.path.isdir(out_dir):
    os.makedirs(out_dir)

monitor = ConsoleTaskMonitor()
fm = currentProgram.getFunctionManager()
sm = currentProgram.getSymbolTable()
rm = currentProgram.getReferenceManager()
listing = currentProgram.getListing()

try:
    decomp_available = True
    from ghidra.app.decompiler import DecompInterface
    decomp = DecompInterface()
    decomp.openProgram(currentProgram)
except Exception:
    decomp_available = False
    decomp = None

try:
    from ghidra.app.decompiler.component import DecompilerUtils
except Exception:
    pass


def addr_text(addr):
    return "0x" + addr.toString().upper()


def fn_key(fn):
    return addr_text(fn.getEntryPoint())


def get_function_at_or_containing(addr):
    fn = fm.getFunctionAt(addr)
    if fn is None:
        fn = fm.getFunctionContaining(addr)
    return fn


def symbols_named(name):
    out = []
    syms = sm.getSymbols(name)
    while syms.hasNext():
        out.append(syms.next())
    return out


def refs_to_symbol(sym):
    refs = []
    it = rm.getReferencesTo(sym.getAddress())
    while it.hasNext():
        refs.append(it.next())
    return refs


def callers_of(fn):
    callers = []
    refs = rm.getReferencesTo(fn.getEntryPoint())
    while refs.hasNext():
        ref = refs.next()
        f = get_function_at_or_containing(ref.getFromAddress())
        if f is not None:
            callers.append(f)
    return callers


def callees_of(fn):
    callees = []
    body = fn.getBody()
    ins = listing.getInstructions(body, True)
    while ins.hasNext():
        inst = ins.next()
        flows = inst.getFlows()
        if flows:
            for a in flows:
                f = get_function_at_or_containing(a)
                if f is not None:
                    callees.append(f)
    return callees


def walk(seed_fns, direction):
    seen = {}
    queue = []
    for fn in seed_fns:
        if fn is not None:
            queue.append((fn, 0))
    while queue:
        fn, depth = queue.pop(0)
        key = fn_key(fn)
        if key in seen and seen[key] <= depth:
            continue
        seen[key] = depth
        if depth >= MAX_DEPTH:
            continue
        next_fns = callers_of(fn) if direction == "callers" else callees_of(fn)
        for nxt in next_fns:
            queue.append((nxt, depth + 1))
    return seen


def contains_anchor(fn):
    body = fn.getBody()
    text = str(fn.getName()) + " " + addr_text(fn.getEntryPoint())
    ins = listing.getInstructions(body, True)
    while ins.hasNext():
        inst = ins.next()
        text += " " + addr_text(inst.getAddress()) + " " + inst.toString()
    hits = []
    for a in VM_ANCHORS:
        if a.upper().replace("0X", "0x") in text or a[2:].upper() in text.upper():
            hits.append(a)
    return hits


def disasm_text(fn):
    lines = []
    ins = listing.getInstructions(fn.getBody(), True)
    while ins.hasNext():
        inst = ins.next()
        lines.append("%s %s" % (addr_text(inst.getAddress()), inst.toString()))
    return "\n".join(lines)


def pcode_text(fn):
    lines = []
    ins = listing.getInstructions(fn.getBody(), True)
    while ins.hasNext():
        inst = ins.next()
        ops = inst.getPcode()
        for op in ops:
            lines.append("%s %s" % (addr_text(inst.getAddress()), op.toString()))
    return "\n".join(lines)


def decompile_text(fn):
    if not decomp_available:
        return ""
    try:
        res = decomp.decompileFunction(fn, 60, monitor)
        if res and res.decompileCompleted():
            return res.getDecompiledFunction().getC()
    except Exception as exc:
        return "DECOMPILE_ERROR: %s" % exc
    return ""


def write_hints(fn, pcode, disasm):
    hints = []
    combined = (pcode + "\n" + disasm).splitlines()
    for line in combined:
        up = line.upper()
        if "STORE" in up or re.search(r"\bMOV\s+\w+ PTR \[", up):
            pattern = "write"
            if "QWORD" in up or ":8" in up or "(8)" in up:
                pattern = "8-byte-write"
            if "INT_ZEXT" in up or "INT_ADD" in up or "INT_XOR" in up or "INT_MULT" in up:
                pattern += "+key-arithmetic-nearby"
            hints.append({
                "function": fn.getName(),
                "function_entry": fn_key(fn),
                "pattern": pattern,
                "line": line[:240]
            })
    return hints

# Seed from imports and external refs.
seed_functions = {}
import_refs = []
for name in NETWORK_NAMES:
    for sym in symbols_named(name):
        refs = refs_to_symbol(sym)
        for ref in refs:
            fn = get_function_at_or_containing(ref.getFromAddress())
            import_refs.append({
                "api": name,
                "symbol_address": addr_text(sym.getAddress()),
                "from_address": addr_text(ref.getFromAddress()),
                "function": fn.getName() if fn else "",
                "function_entry": fn_key(fn) if fn else ""
            })
            if fn is not None:
                seed_functions[fn_key(fn)] = fn

# Also seed known VM-anchor containing functions, if present.
for anchor in VM_ANCHORS:
    try:
        addr = toAddr(anchor)
        fn = get_function_at_or_containing(addr)
        if fn is not None:
            seed_functions[fn_key(fn)] = fn
    except Exception:
        pass

all_targets = {}
caller_depth = walk(seed_functions.values(), "callers")
callee_depth = walk(seed_functions.values(), "callees")
for fn in seed_functions.values():
    all_targets[fn_key(fn)] = {"fn": fn, "seed": True, "caller_depth": 0, "callee_depth": 0}
for key, depth in caller_depth.items():
    fn = get_function_at_or_containing(toAddr(key))
    all_targets.setdefault(key, {"fn": fn, "seed": False})["caller_depth"] = depth
for key, depth in callee_depth.items():
    fn = get_function_at_or_containing(toAddr(key))
    all_targets.setdefault(key, {"fn": fn, "seed": False})["callee_depth"] = depth

call_edges = []
function_rows = []
write_rows = []
for key in sorted(all_targets.keys()):
    item = all_targets[key]
    fn = item["fn"]
    if fn is None:
        continue
    dtext = disasm_text(fn)
    ptext = pcode_text(fn)
    ctext = decompile_text(fn)
    anchors = contains_anchor(fn)
    base = "%s_%s" % (key.replace("0x", ""), re.sub(r"[^A-Za-z0-9_]+", "_", fn.getName())[:80])
    open(os.path.join(out_dir, base + ".disasm.txt"), "w").write(dtext)
    open(os.path.join(out_dir, base + ".pcode.txt"), "w").write(ptext)
    open(os.path.join(out_dir, base + ".decomp.txt"), "w").write(ctext)
    write_rows.extend(write_hints(fn, ptext, dtext))
    function_rows.append({
        "entry": key,
        "name": fn.getName(),
        "seed": bool(item.get("seed", False)),
        "caller_depth": item.get("caller_depth", ""),
        "callee_depth": item.get("callee_depth", ""),
        "vm_anchor_hits": ";".join(anchors),
        "disasm_file": base + ".disasm.txt",
        "pcode_file": base + ".pcode.txt",
        "decomp_file": base + ".decomp.txt"
    })
    for callee in callees_of(fn):
        call_edges.append({"from_entry": key, "from_name": fn.getName(), "to_entry": fn_key(callee), "to_name": callee.getName()})

open(os.path.join(out_dir, "import_refs.json"), "w").write(json.dumps(import_refs, indent=2, sort_keys=True))
open(os.path.join(out_dir, "candidate_functions.json"), "w").write(json.dumps(function_rows, indent=2, sort_keys=True))
open(os.path.join(out_dir, "call_edges.json"), "w").write(json.dumps(call_edges, indent=2, sort_keys=True))
open(os.path.join(out_dir, "write_hints.json"), "w").write(json.dumps(write_rows, indent=2, sort_keys=True))
open(os.path.join(out_dir, "export_manifest.json"), "w").write(json.dumps({
    "program": currentProgram.getName(),
    "network_names": NETWORK_NAMES,
    "vm_anchors": VM_ANCHORS,
    "max_depth": MAX_DEPTH,
    "candidate_function_count": len(function_rows),
    "write_hint_count": len(write_rows),
    "note": "Local-only p-code/disassembly export. Summarize before committing."
}, indent=2, sort_keys=True))
print("Pass622 S2C receive export wrote %d functions and %d write hints to %s" % (len(function_rows), len(write_rows), out_dir))
